vendor/noise_handshake/src/handshake.rs

```diff
@@ -0,0 +1,230 @@
+use futures::*;
+use snow::Session::{self, Handshake};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+const NOISE_MAX_SIZE: usize = 65535;
+
+#[derive(DebugStub)]
+pub enum Step {
+    Read {
+        #[debug_stub = "encrypted buffer"]
+        enc_buffer: [u8; NOISE_MAX_SIZE],
+        len: usize,
+    },
+    Write {
+        #[debug_stub = "buffer"]
+        buffer: [u8; NOISE_MAX_SIZE],
+        remaining_bytes: usize,
+        len: usize,
+    },
+}
+
+impl Step {
+    fn read() -> Self {
+        Step::Read {
+            enc_buffer: [0u8; NOISE_MAX_SIZE],
+            len: 0,
+        }
+    }
+
+    fn write() -> Self {
+        Step::Write {
+            buffer: [0u8; NOISE_MAX_SIZE],
+            remaining_bytes: 0,
+            len: 0,
+        }
+    }
+}
+
+pub trait InitHandshake {
+    fn init_hand_shake<IO>(self, io: IO) -> NoiseHandshake<IO>
+    where
+        IO: AsyncRead + AsyncWrite;
+}
+
+impl InitHandshake for Session {
+    fn init_hand_shake<IO>(self, io: IO) -> NoiseHandshake<IO>
+    where
+        IO: AsyncRead + AsyncWrite,
+    {
+        match self {
+            Handshake(ref handshake_state) => NoiseHandshake {
+                next: if handshake_state.is_initiator() {
+                    Step::write()
+                } else {
+                    Step::read()
+                },
+                noise: Some(self),
+                io: Some(io),
+            },
+            _ => panic!(
+                "Noise Session in incorrect state, you should init before starting handshake"
+            ),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct NoiseHandshake<IO: AsyncRead + AsyncWrite> {
+    next: Step,
+    noise: Option<Session>,
+    io: Option<IO>,
+}
+
+impl<IO: AsyncRead + AsyncWrite> Future for NoiseHandshake<IO> {
+    type Item = (Session, IO);
+    // TODO: probably a custom error when removing unwraps
+    type Error = std::io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self {
+            Self {
+                noise: Some(noise),
+                io: Some(_),
+                ..
+            } if noise.is_handshake_finished() => {
+                let noise = self.noise.take().expect("We know it's a Some");
+                let noise = noise.into_transport_mode().unwrap();
+                let io = self.io.take().expect("We know it's a Some");
+                Ok(Async::Ready((noise, io)))
+            }
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(ref mut io),
+                next:
+                    Step::Write {
+                        mut buffer,
+                        mut remaining_bytes,
+                        mut len,
+                    },
+            } if remaining_bytes == 0 => {
+                len = noise.write_message(&[], &mut buffer).unwrap();
+
+                let mut num_bytes_written = 0;
```

Comment:
:wrench: Remove the whole block until `self.poll` and let the block below handle that :)